### ISSUE TITLE
Mqtt pubsub components fails to read messages on startup when multiple subscriptions configured

### ISSUE THREAD

**Issue by bwojdyla** [dapr/components-contrib#2264]
## Expected Behavior

Messages that are persisted on queue while dapr sidecar was down should be processed after dapr sidecar is started.

## Actual Behavior

After darp sidecar is started messages which are one queue are not processed.

## Steps to Reproduce the Problem

Configure 2 subscribtions for mqtt component:
```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: testapp-pubsub
spec:
  type: pubsub.mqtt
  version: v1
  metadata:
  - name: url
    value: "abc"
  - name: qos
    value: 1
  - name: retain
    value: "false"
  - name: cleanSession
    value: "false"
  - name: backOffMaxRetries
    value: "5"
```

Subscription 1

```
apiVersion: dapr.io/v1alpha1
kind: Subscription
metadata:
  name: subscription1 
spec:
  pubsubname: testapp-pubsub
  topic: topic1
  route: /api/route1
```

Subscription 2

```
apiVersion: dapr.io/v1alpha1
kind: Subscription
metadata:
  name: subscription2
spec:
  pubsubname: testapp-pubsub
  topic: topic2
  route: /api/route2
```
1. Start dapr + application for first time so that it creates queue
2. Stop dapr
3. Send at least 2 messages for topic1
4. Start dapr for second time
5. You should see similar behavior:
```
time="2022-11-08T16:27:09.771556034Z" level=info msg="app is subscribed to the following topics: [topic1 topic2] through pubsub=testapp-pubsub" app_id=test.testapp.api instance=536c465db880 scope=dapr.runtime type=log ver=1.9.3
time="2022-11-08T16:27:09.771606899Z" level=debug msg="subscribing to topic='topic1' on pubsub='testapp-pubsub'" app_id=test.testapp.api instance=536c465db880 scope=dapr.runtime type=log ver=1.9.3
time="2022-11-08T16:27:09.771680668Z" level=info msg="initializing the subscriber" app_id=test.testapp.api instance=536c465db880 scope=dapr.contrib type=log ver=1.9.3
time="2022-11-08T16:27:09.777840715Z" level=debug msg="subscribing to topic='topic2' on pubsub='testapp-pubsub'" app_id=test.testapp.api instance=536c465db880 scope=dapr.runtime type=log ver=1.9.3
time="2022-11-08T16:27:09.777891519Z" level=debug msg="Processing MQTT message topic1#130 (retained=false)" app_id=test.testapp.api instance=536c465db880 scope=dapr.contrib type=log ver=1.9.3
time="2022-11-08T16:27:09.77789804Z" level=info msg="re-initializing the subscriber" app_id=test.testapp.api instance=536c465db880 scope=dapr.contrib type=log ver=1.9.3
time="2022-11-08T16:27:09.798402511Z" level=debug msg="Done processing MQTT message topic1#130; sending ACK" app_id=test.testapp.api instance=536c465db880 scope=dapr.contrib type=log ver=1.9.3
time="2022-11-08T16:27:39.689842717Z" level=debug msg="Refreshing all mDNS addresses." app_id=test.testapp.api instance=536c465db880 scope=dapr.contrib type=log ver=1.9.3
time="2022-11-08T16:27:39.689979132Z" level=debug msg="no mDNS apps to refresh." app_id=test.testapp.api instance=536c465db880 scope=dapr.contrib type=log ver=1.9.3
time="2022-11-08T16:28:09.690739406Z" level=debug msg="Refreshing all mDNS addresses." app_id=test.testapp.api instance=536c465db880 scope=dapr.contrib type=log ver=1.9.3
time="2022-11-08T16:28:09.690850391Z" level=debug msg="no mDNS apps to refresh." app_id=test.testapp.api instance=536c465db880 scope=dapr.contrib type=log ver=1.9.3
time="2022-11-08T16:28:39.691071766Z" level=debug msg="Refreshing all mDNS addresses." app_id=test.testapp.api instance=536c465db880 scope=dapr.contrib type=log ver=1.9.3
time="2022-11-08T16:28:39.691134361Z" level=debug msg="no mDNS apps to refresh." app_id=test.testapp.api instance=536c465db880 scope=dapr.contrib type=log ver=1.9.3
```

1. Subscribing for topic1
2. Subscribing for topic2
3. Only one message is being processed but isn't removed from queue
4. No more messages which are on queue are processed
5. Also messages that are being send after start are not processed

Notice that:
- if there is less than 2 messages on queue (0 or 1) mqtt component works fine and messages are processed
- in scenario above if there were messages for topic2 on queue then it would work fine

It seems that adding second subscription causes race condition.
```
time="2022-11-08T16:27:09.77789804Z" level=info msg="re-initializing the subscriber" app_id=test.testapp.api instance=536c465db880 scope=dapr.contrib type=log ver=1.9.3
```

**Comment by akhilac1** [comment]
Ack'd the issue @bwojdyla. Will ping back once we replicate this behaviour. @pravinpushkar @addjuarez

**Comment by berndverst** [comment]
FYI @ItalyPaleAle

**Comment by pravinpushkar** [comment]
@bwojdyla I could not reproduce the issue, it seems to work fine for me. Please find below details steps and logs. Please let us know if there needs to any modification to the steps below -

- Start subscriber app to subscribe to 2 topics A and B. Took above shared yaml files but modified the topic names to keep simple.
- Start publisher app and send 5 messages only to topic A.
- Stop subscriber app but keep sending data to topic A. While subscribe app was down, publisher sent 3 more messages.
- Bring subscriber app up again.
- [**Result**]All message that was sent was received.
           1.  Initially when both apps were up. All Messages sent were received by subscriber on topic A.
           2. All messages that were sent when subscriber went down were received when subscriber came online again.
           3. All message that were sent thereafter were being received normally.

 ### Logs - 

#### 1. Subscriber and publisher both were up initially - 

**Publisher Logs**

INFO[0000] app is subscribed to the following topics: [B A] through pubsub=orderpubsub  app_id=checkout-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.runtime type=log ver=edge
INFO[0000] initializing the subscriber                   app_id=checkout-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.contrib type=log ver=edge
== APP == Published data to Topic A: {"orderId_A":1}
INFO[0000] re-initializing the subscriber                app_id=checkout-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.contrib type=log ver=edge
INFO[0000] placement tables updated, version: 0          app_id=checkout-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.runtime.actor.internal.placement type=log ver=edge
INFO[0000] dapr initialized. Status: Running. Init Elapsed 30ms  app_id=checkout-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.runtime type=log ver=edge
ℹ️  Updating metadata for app command: node .
✅  You're up and running! Both Dapr and your app logs will appear here.

== APP == Published data to Topic A: {"orderId_A":2}
== APP == Published data to Topic A: {"orderId_A":3}
== APP == Published data to Topic A: {"orderId_A":4}
== APP == Published data to Topic A: {"orderId_A":5}

**Subscriber Logs**

== APP == Subscriber received message on Topic A: { orderId_A: 1 }
== APP == Subscriber received message on Topic A: { orderId_A: 2 }
== APP == Subscriber received message on Topic A: { orderId_A: 2 }
== APP == Subscriber received message on Topic A: { orderId_A: 3 }
== APP == Subscriber received message on Topic A: { orderId_A: 3 }
== APP == Subscriber received message on Topic A: { orderId_A: 4 }
== APP == Subscriber received message on Topic A: { orderId_A: 4 }
== APP == Subscriber received message on Topic A: { orderId_A: 5 }
== APP == Subscriber received message on Topic A: { orderId_A: 5 }

#### 2. Put down subscriber while publisher keeps on sending - 
**Publisher log -** 
== APP == Published data to Topic A: {"orderId_A":6}
== APP == Published data to Topic A: {"orderId_A”:7}
== APP == Published data to Topic A: {"orderId_A”:8}

**Bring subscriber app online - 
Subscriber logs -** 
INFO[0000] app is subscribed to the following topics: [B A] through pubsub=orderpubsub  app_id=order-processor-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.runtime type=log ver=edge
INFO[0000] initializing the subscriber                   app_id=order-processor-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.contrib type=log ver=edge
INFO[0000] re-initializing the subscriber                app_id=order-processor-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.contrib type=log ver=edge
INFO[0000] placement tables updated, version: 0          app_id=order-processor-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.runtime.actor.internal.placement type=log ver=edge
== APP == Subscriber received message on Topic A: { orderId_A: 7 }
== APP == Subscriber received message on Topic A: { orderId_A: 8 }
== APP == Subscriber received message on Topic A: { orderId_A: 6 }


#### 3. Later logs - 

Publisher - 

== APP == Published data to Topic A: {"orderId_A":9}
== APP == Published data to Topic A: {"orderId_A":10}
== APP == Published data to Topic A: {"orderId_A":11}
== APP == Published data to Topic A: {"orderId_A":12}
== APP == Published data to Topic A: {"orderId_A":13}
== APP == Published data to Topic A: {"orderId_A":14}


Subscriber - 
== APP == Subscriber received message on Topic A: { orderId_A: 9 }
== APP == Subscriber received message on Topic A: { orderId_A: 10 }
== APP == Subscriber received message on Topic A: { orderId_A: 11 }
== APP == Subscriber received message on Topic A: { orderId_A: 12 }
== APP == Subscriber received message on Topic A: { orderId_A: 13 }
== APP == Subscriber received message on Topic A: { orderId_A: 14 }

**Comment by bwojdyla** [comment]
@pravinpushkar in logs you provided there is no information which topic was subscribed first
```
INFO[0000] app is subscribed to the following topics: [B A] through pubsub=orderpubsub app_id=order-processor-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.runtime type=log ver=edge
INFO[0000] initializing the subscriber app_id=order-processor-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.contrib type=log ver=edge
INFO[0000] re-initializing the subscriber app_id=order-processor-http instance=Pravins-MBP.guest.corp.microsoft.com scope=dapr.contrib type=log ver=edge
```
`re-initializing` happens but to determine which topic was subscribed first you need to enable debug logs.

In your case you should be able to reproduce issue if component subscribes to topic A first and then to topic B.
Also try to add some waiting in subscriber's message handler. 1s was enough for me. But I guess more is better.

**Comment by pravinpushkar** [comment]
@bwojdyla 

> component subscribes to topic A first and then to topic B

The subscription happens through yaml files and AFAIK there is no way to specify which topic gets subscribed first.
But, to simulate the behaviour, this is what I did - 

1. Kept only one subscription yaml i.e for topic 'A'
2. started subscriber app and publisher app
3. Publisher publishes 5 messages and all were received successfully at subscriber app.
4. stopped subscriber app and let publisher publish 3 messages.
5. copy the other subscription yaml for topic 'B' also in component folder.
6. start the subscriber app
7. All messages received successfully and also it is receiving the currently getting published items.

I also tried below after step 6 and this also seemed to work fine.
7. start publishing to topic B also along with topic A.
8. All previous messages that were published to topic A while subscriber was down received along with currently getting published items on both topics A and B.

Note - I tested this with latest dapr runtime 1.9.3 as well as edge version and I observed the same behavior.

**Comment by pravinpushkar** [comment]
@bwojdyla Please see the below logs. 

Initial logs. Messages from 1 to 6 on topic A get delivered before shutting down.

INFO[0000] **app is subscribed to the following topics: [B A]** through pubsub=orderpubsub  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
DEBU[0000] **subscribing to topic='A'** on pubsub='orderpubsub'  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
INFO[0000] initializing the subscriber                   app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
DEBU[0000] **subscribing to topic='B'** on pubsub='orderpubsub'  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
INFO[0000] re-initializing the subscriber                app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
INFO[0000] dapr initialized. Status: Running. Init Elapsed 689ms  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
DEBU[0000] established connection to placement service at dns:///localhost:50005  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor.internal.placement type=log ver=1.9.3
DEBU[0000] placement order received: lock                app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor.internal.placement type=log ver=1.9.3
DEBU[0000] placement order received: update              app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor.internal.placement type=log ver=1.9.3
INFO[0000] placement tables updated, version: 0          app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor.internal.placement type=log ver=1.9.3
DEBU[0000] placement order received: unlock              app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor.internal.placement type=log ver=1.9.3
INFO[0000] HTTP API Called: PUT /v1.0/metadata/cliPID    app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.http-info type=log useragent=Go-http-client/1.1 ver=1.9.3
ℹ️  Updating metadata for app command: node .
INFO[0000] HTTP API Called: PUT /v1.0/metadata/appCommand  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.http-info type=log useragent=Go-http-client/1.1 ver=1.9.3
✅  You're up and running! Both Dapr and your app logs will appear here.

INFO[0021] HTTP API Called: GET /v1.0/metadata           app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.http-info type=log useragent=Go-http-client/1.1 ver=1.9.3
DEBU[0021] Processing MQTT message A#1 (retained=false)  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
== APP == Subscriber received message on Topic A: { orderId_A: 1 }
== APP == Subscriber received message on Topic A: { orderId_A: 1 }
DEBU[0021] skipping status check due to error parsing result from pub/sub event 7a749b64-a4bb-4f80-9f49-aa01622453e9  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
DEBU[0021] Done processing MQTT message A#1; sending ACK  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
DEBU[0026] Processing MQTT message A#2 (retained=false)  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
== APP == Subscriber received message on Topic A: { orderId_A: 2 }
DEBU[0026] skipping status check due to error parsing result from pub/sub event 829ef0c1-c3cc-4266-bf33-b952612f2404  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
DEBU[0026] Done processing MQTT message A#2; sending ACK  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
== APP == Subscriber received message on Topic A: { orderId_A: 2 }
DEBU[0030] Refreshing all mDNS addresses.                app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
DEBU[0030] no mDNS apps to refresh.                      app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
== APP == Subscriber received message on Topic A: { orderId_A: 3 }
DEBU[0031] Processing MQTT message A#3 (retained=false)  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
== APP == Subscriber received message on Topic A: { orderId_A: 3 }
DEBU[0031] skipping status check due to error parsing result from pub/sub event b9430e81-eca1-453f-b735-89689098e5a0  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
DEBU[0031] Done processing MQTT message A#3; sending ACK  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
DEBU[0036] Processing MQTT message A#4 (retained=false)  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
== APP == Subscriber received message on Topic A: { orderId_A: 4 }
== APP == Subscriber received message on Topic A: { orderId_A: 4 }
DEBU[0036] skipping status check due to error parsing result from pub/sub event b39aafb2-d37a-4c9a-81f2-f47cf54c59b0  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
DEBU[0036] Done processing MQTT message A#4; sending ACK  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
DEBU[0041] Processing MQTT message A#5 (retained=false)  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
== APP == Subscriber received message on Topic A: { orderId_A: 5 }
== APP == Subscriber received message on Topic A: { orderId_A: 5 }
DEBU[0041] skipping status check due to error parsing result from pub/sub event 2cf73516-4b5b-4250-8eca-4c5f3cb3a421  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
DEBU[0041] Done processing MQTT message A#5; sending ACK  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
DEBU[0046] Processing MQTT message A#6 (retained=false)  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
== APP == Subscriber received message on Topic A: { orderId_A: 6 }
== APP == Subscriber received message on Topic A: { orderId_A: 6 }
DEBU[0046] skipping status check due to error parsing result from pub/sub event e4713306-c26a-4901-8af4-a6a9d3e0bf9c  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
DEBU[0046] Done processing MQTT message A#6; sending ACK  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
^CDEBU[0049] received SIGTERM signal, shutting down...     app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
ℹ️
terminated signal received: shutting down

--------------------------------------------------------------------








These logs are captured after restarting the subscriber and topic A had message from 7-12 on the queue(**_see bold and highlighted_**), Which get delivered as soon as the subscriber is back online.

INFO[0000] internal gRPC server is running on port 50318  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
INFO[0000] application protocol: http. waiting on port 5001.  This will block until the app is listening on that port.  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
_**== APP == Subscriber received message on Topic A: { orderId_A: 8 }
== APP == Subscriber received message on Topic A: { orderId_A: 11 }
== APP == Subscriber received message on Topic A: { orderId_A: 12 }
== APP == Subscriber received message on Topic A: { orderId_A: 7 }
== APP == Subscriber received message on Topic A: { orderId_A: 10 }
== APP == Subscriber received message on Topic A: { orderId_A: 9 }**_
INFO[0000] application discovered on port 5001           app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
INFO[0000] application configuration loaded              app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
INFO[0000] actors: state store is not configured - this is okay for clients but services with hosted actors will fail to initialize!  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
INFO[0000] actor runtime started. actor idle timeout: 1h0m0s. actor scan interval: 30s  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor type=log ver=1.9.3
DEBU[0000] try to connect to placement service: dns:///localhost:50005  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor.internal.placement type=log ver=1.9.3
DEBU[0000] user app did not subscribe to any topic       app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
DEBU[0000] app responded with subscriptions []           app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
INFO[0000]  **app is subscribed to the following topics: [A B]** through pubsub=orderpubsub  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
DEBU[0000]  **subscribing to topic='A'** on pubsub='orderpubsub'  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
INFO[0000] initializing the subscriber                   app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
DEBU[0000] established connection to placement service at dns:///localhost:50005  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor.internal.placement type=log ver=1.9.3
DEBU[0000] placement order received: lock                app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor.internal.placement type=log ver=1.9.3
DEBU[0000] placement order received: update              app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor.internal.placement type=log ver=1.9.3
INFO[0000] placement tables updated, version: 0          app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor.internal.placement type=log ver=1.9.3
DEBU[0000] placement order received: unlock              app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.actor.internal.placement type=log ver=1.9.3
INFO[0000] HTTP API Called: PUT /v1.0/metadata/cliPID    app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.http-info type=log useragent=Go-http-client/1.1 ver=1.9.3
ℹ️  Updating metadata for app command: node .
INFO[0000] HTTP API Called: PUT /v1.0/metadata/appCommand  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime.http-info type=log useragent=Go-http-client/1.1 ver=1.9.3
✅  You're up and running! Both Dapr and your app logs will appear here.

== APP == Subscriber received message on Topic A: { orderId_A: 13 }
== APP == Subscriber received message on Topic A: { orderId_A: 14 }
== APP == Subscriber received message on Topic A: { orderId_A: 15 }
== APP == Subscriber received message on Topic A: { orderId_A: 16 }
== APP == Subscriber received message on Topic A: { orderId_A: 17 }
== APP == Subscriber received message on Topic A: { orderId_A: 18 }
DEBU[0030] Refreshing all mDNS addresses.                app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
DEBU[0030] no mDNS apps to refresh.                      app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
DEBU[0030] **subscribing to topic='B'** on pubsub='orderpubsub'  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
INFO[0030] re-initializing the subscriber                app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
DEBU[0030] Processing MQTT message A#7 (retained=false)  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
**_== APP == Subscriber received message on Topic A: { orderId_A: 7 }_**
DEBU[0030] skipping status check due to error parsing result from pub/sub event 2c4851b4-dc9a-40bc-9114-7a50eace5f42  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.runtime type=log ver=1.9.3
DEBU[0030] Done processing MQTT message A#7; sending ACK  app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
== APP == Subscriber received message on Topic A: { orderId_A: 19 }
== APP == Subscriber received message on Topic A: { orderId_A: 20 }
== APP == Subscriber received message on Topic A: { orderId_A: 21 }
== APP == Subscriber received message on Topic A: { orderId_A: 22 }
== APP == Subscriber received message on Topic A: { orderId_A: 23 }
== APP == Subscriber received message on Topic A: { orderId_A: 24 }
DEBU[0060] Refreshing all mDNS addresses.                app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
DEBU[0060] no mDNS apps to refresh.                      app_id=order-processor-http instance=Pravins-MacBook-Pro.local scope=dapr.contrib type=log ver=1.9.3
== APP == Subscriber received message on Topic A: { orderId_A: 25 }
== APP == Subscriber received message on Topic A: { orderId_A: 26 }
== APP == Subscriber received message on Topic A: { orderId_A: 27 }

**Comment by bwojdyla** [comment]
I'm using daprio/daprd:1.9.3 configured according:
https://docs.dapr.io/operations/hosting/self-hosted/self-hosted-with-docker/#run-using-docker-compose

I think you reproduced it.
For each message you should have following logs:
```
Processing MQTT message A#<no>
Done processing MQTT message A#<no>
```

And you have those entries for messages 1-6 before restart.
After restarting subscriber there are entries only for single message i.e. A#7.

In my case only single message reaches application and none of those are removed from queue after restart.

**Comment by pravinpushkar** [comment]
@bwojdyla 
I see. So in your case you are not even receiving the message (after receiving the 1st message on restarting the subscriber) on the subscriber app ?
What my observation is I can receive the messages but agree that ACK is not present after the 1st message on restart.

Which MQTT broker you are using ?

**Comment by bwojdyla** [comment]
> I see. So in your case you are not even receiving the message (after receiving the 1st message on restarting the subscriber) on the subscriber app ?

That's correct. I receive only first message on the subscriber app.

> Which MQTT broker you are using ?

Solace
https://docs.solace.com/API/MQTT/MQTT-get-started.htm

>  By any chance, you have a dev setup for Dapr?

I don't have. And I don't have experience with golang.

**Comment by pravinpushkar** [comment]
@bwojdyla ok. got it. Is it possible to try with other broker just for testing like mosquito. You can just start mosquito with this command - `docker run -d -p 1883:1883 -p 9001:9001 --name mqtt eclipse-mosquitto:1.6`
and the component yaml file will be something like this - 

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: orderpubsub
spec:
  type: pubsub.mqtt
  version: v1
  metadata:
  - name: url
    value: tcp://localhost:1883
  - name: authType
    value: none
  - name: qos
    value: 1
  - name: retain
    value: "false"
  - name: cleanSession
    value: "false"
  - name: backOffMaxRetries
    value: "5"

```

**Comment by bwojdyla** [comment]
@pravinpushkar 
I tested it with mosquitto and had same behavior on version daprio/daprd:1.9.3. So after restart only first message is delivered to app and none are acknowledged.

But after switching to daprio/daprd:edge messages are delivered and acknowledged. Both for solace and mosquitto.

Other issue is that messages are delivered twice on edge.
My guess is that it is because of client being reconnected.
Here are mosquitto logs:
```
1668592366: New connection from 172.18.0.4 on port 1883.
1668592366: New client connected from 172.18.0.4 as app-producer (p2, c0, k30).
1668592366: New connection from 172.18.0.4 on port 1883.
1668592366: New client connected from 172.18.0.4 as app-consumer (p2, c0, k30).
1668592366: Client app-consumer disconnected.
1668592366: New connection from 172.18.0.4 on port 1883.
1668592366: New client connected from 172.18.0.4 as app-consumer (p2, c0, k30).
```
Client will be disconnected once if there are 2 topics, twice if there are 3 topics, etc.

Should I report it in new issue for clarity? Or this is expected behavior?

**Comment by pravinpushkar** [comment]
> I tested it with mosquitto and had same behavior on version daprio/daprd:1.9.3. So after restart only first message is delivered to app and none are acknowledged

Yeah, I think you are right. I am also able to reproduce it now on 1.9.3.

> Other issue is that messages are delivered twice on edge.

I guess you meant the pending messages only(when subscriber was down) and that too not all are getting delivered twice, right ?

**Comment by pravinpushkar** [comment]
Summary -  
- Issue is reproducible with 1.9.3 and needs a fix.
- Issue does not exists with dapr edge version however the pending messages are getting delivered multiple times.

**Comment by pravinpushkar** [comment]
@bwojdyla Thanks for reporting this and sharing inputs and logs. We are in the process of releasing a hotfix for this.

Feel free to create a separate issue regarding the multiple delivery of pending messages. At the moment this looks fine due to "at least one" semantic for message delivery. But if you feel otherwise please create a separate issue for discussion.

**Comment by pravinpushkar** [comment]
@bwojdyla Feel free to close this.
